### PR TITLE
NON-280 Add service to find gangs with non-associations

### PR DESCRIPTION
**Pull request by Mjwillis**
Work in progress.  

Will add controllers and API with integration tests next,

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/Gang.kt [hunk 1]
```diff
@@ -44,8 +44,19 @@ class Gang(
...
 
+  @OneToMany(mappedBy = "primaryGang", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
+  val nonAssociationsPrimary: List<GangNonAssociation> = listOf(),
+
+  @OneToMany(mappedBy = "secondaryGang", fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
+  val nonAssociationsSecondary: List<GangNonAssociation> = listOf(),
+
 ) {
 
+  fun getNonAssociations(): List<Pair<Gang, NonAssociationReason>> =
```
Comment (Mjwillis) [inline src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/Gang.kt:55]: This returns a list of Gangs and that should not be associated with the current gang and the reason for the non-association

File: src/test/java/uk/gov/justice/hmpps/prison/service/GangServiceIntTest.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    )
+
+    TestTransaction.flagForCommit()
+    TestTransaction.end()
+  }
+
+  @Test
+  fun getNonAssociatesInGangs() {
+    val gangNas = gangService.getNonAssociatesInGangs("A1234AE")
+    Assertions.assertThat(gangNas.gangNonAssociations).hasSize(2)
```
Comment (Mjwillis) [inline src/test/java/uk/gov/justice/hmpps/prison/service/GangServiceIntTest.kt:129]: Will add more here shortly

File: src/main/java/uk/gov/justice/hmpps/prison/service/GangService.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+@Schema(description = "Gang Summary")
+data class GangSummary(
+  @Schema(description = "The code assigned for the gang", example = "A_GANG")
+  val code: String,
+  @Schema(description = "The name of the gang", example = "A New Gang")
+  val name: String,
+  @Schema(description = "Information about this member within the gang", example = "Leader of gang")
+  val comment: String?,
+  @Schema(description = "Number of members in this gang", example = "15")
+  val memberOfMembers: Long,
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/hmpps/prison/service/GangService.kt:85]: typo in name?

Comment (Mjwillis) [reply src/main/java/uk/gov/justice/hmpps/prison/service/GangService.kt]: fixed

**Review by steverendell — APPROVED** [review]
lgtm

**Pull request merged**